Misc fixes for applying IPv6  static address.

cathay4t opened the pull request:
~~* Raise exception when user try to flip IPv6 status(enable/disable).~~
    Other Design not been chose:
     * Option A: Use sysctl to disable/enable IPv6.
        * Pros:
            * Works as expected for IPv6 disable and enable.
        * Cons:
            * Need to parse multiple localtion of sysctl.d(/run, /etc,
              /usr/lib, /user/local, ....)
            * Need to handle when user disable/enable IPv6 using their own
              file of sysctl.conf.
            * Docker does not allow sysctl inside.
     * Option B: Hide IPv6 link local address.
        * Pros:
            * No hack required to support IPv6 enable/disable via
              NetworkManager.
        * Cons
            * We does not showing the precise current network state.
 * Ordering the IP addresses before verify comparison of `netapplier.apply()`.
~~ * Raise `MissingIPv6LinkLocalAddressError` when no IPv6 link local address.~~
 * Ignore IPv6 link-local address when `netapplier.apply()`.
 * I don't add the test case for IPv6 yet. Just waiting other PR get merged.

cathay4t commented:
@tyll Can you update the docker image as I add `ipaddress` denpendency. Thank you.

tyll commented on libnmstate/netapplier.py:
```diff
@@ -313,12 +315,20 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
                 'desired: {}\ncurrent: {}'.format(iface_dstate, iface_cstate))
 
 
+def _sort_ip_addresses(iface):
```
I remember that @thom311 told me that the order of IP addresses matters but I forgot for what exactly. Maybe for the default IP address that is used for outgoing connections via that interface.

EdDev replied:
On setting it probably matters who is the primary. If this is so, then only the 1st one is important, all the others order is not important.
We can use a metadata on input to mark the primary, then convert the list to a set. On reporting, we can convert the set to a list and place the primary first.

cathay4t replied:
In Networkmanager source code, the `src/nm-ip4-config.c:_addresses_sort_cmp()` indicates, if no priority defined, for the same network addresss, sort lexically and the first one will be set as primary, others be secondary. So we are doing correct here.

And I will rename the `iface` to `iface_state`.

tyll commented on libnmstate/netapplier.py:
```diff
@@ -312,12 +316,20 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
+    for family in ('ipv4', 'ipv6'):
+        if iface.get(family, {}).get('address', []):
+            iface[family]['address'] = sorted(iface[family]['address'],
+                                              key=itemgetter('ip'))
```
For programmers that do not know functional programming itemgetter is easier to understand - also it has a descriptive name IMHO.

cathay4t replied:
@EdDev I prefer the lamdba way also, but pylint suggest I should use itemgetter instead.

EdDev replied:
@tyll , I think someone did a few steps too much to give such a simple lambda a name, I rarely seen it in use and surprises me more than helps.
@cathay4t , interesting... what is its reasoning?

Maybe I need to learn like in Java all possible options of lambda structures (?!)... if the basics are strong (lambda), we do not need naming for every little thing... it is in front of us to read.

EdDev replied:
Funny thing is, that if you wanted some small deviation like a default value in case it is missing, then you would have to go back to the basics (lambda).

cathay4t commented on libnmstate/nm/iplib.py:
```diff
@@ -0,0 +1,49 @@
+    """
+    a = ipaddress.ip_network(six.u(str(network_a)), strict=False)
+    b = ipaddress.ip_network(six.u(str(network_b)), strict=False)
+    if hasattr(a, 'subnet_of'):
```
Even if I create a psudo `ipaddess` module, I still need another module to provide a function for
subnet calculation, then why two module?

Can you elaborate on the `compat` module? Is that [this module][1]?

[1]: https://python-utilities.readthedocs.io/en/latest/compat.html

EdDev replied:
Ahh sorry, I forgot to give an example: https://github.com/oVirt/vdsm/blob/master/lib/vdsm/common/compat.py

You basically need to inherit in case it is not included and add what you need.
At the place it is used, it should not know you did it.

cathay4t commented on libnmstate/nm/iplib.py:
```diff
@@ -0,0 +1,49 @@
+IPV6_LINK_LOCAL_NETWORK = "fe80::/10"
+
+
+def _is_subnet_of(a, b):
```
As I said in comment, it was copied from Cpython code, I don't intend to make any changes that
the original code in order to save the maintenance effort.

EdDev replied:
I was talking about the order of the functions placed in this file. Please place the functions that call other functions first, such that we can read from to to button, from abstract to details.

cathay4t commented:
> >   * Raise exception when user try to flip IPv6 status(enable/disable).

> I prefer not to go into this area. Either it is always enabled (when available) or disabled because we do not support it.
Raise exception to warn user the reason might be better than failure with `DesiredStateIsNotCurrentError`.

> I do not understand these points. We just need to control /proc/sys/net/ipv6/conf/<iface-name>/disable_ipv6
We need to make it persistent after reboot. Clearly we don't intend to have the nmstate daemon for that.

EdDev commented:
> User might remove link-local address or just replace link-local address to global one. In that case we will get DesiredStateIsNotCurrentError error as link-local will always exists. So I think a new exception explaining the reason could be better.

User cannot remove the default link-local address. Replacing is also not possible, that prefix is reserved.

cathay4t commented:
V2 patch set uploaded.

Changes after V1:

 * New patch to ignore IPv6 link local address when applying state.
 * Removed the two patches doing pre-check on IPv6 link local address.
 * Changed the IP address ordering patch to only do IP address ordering in the verification stage of `netapplier.apply()`. Detail technical detail has been included in commit message.

cathay4t commented:
@EdDev @tyll Please review the V2 patch set.
For the disable ipv6 statue via `/proc`, we need to do that after NM finished its profile activation work as
NM will restore `/proc` `disable_ipv6` file when activating a profile.
This require some works on waiting async actions to be finished, I will submit them in seperate PR.
For current PR, we just:
 * Ignore ip link local address in apply()
 * User will still get `DesiredStateIsNotCurrentError` exception when trying to flip IPv6 state.
 * Ordering IP address internally before verify.

EdDev commented on tox.ini:
```diff
@@ -11,6 +11,7 @@ deps =
     PyGObject
     pytest-cov
     pytest==3.5.1
+    ipaddress
```
In the tox file we add only test env related dependencies, not production (or in the future, build)  ones.
It needs to be placed in the requirements.txt.
You also need to be careful on what version we depend on, as we will need nmstate in RHEL/Centos, so the relevant version should be available there.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -46,15 +52,23 @@ def get_info(active_connection):
         setting_ip.props.method = (
             nmclient.NM.SETTING_IP6_CONFIG_METHOD_MANUAL)
         for address in config['address']:
+            addr_str = '%s/%d' % (address['ip'], address['prefix-length'])
```
Please use format().

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -46,15 +52,23 @@ def get_info(active_connection):
             setting_ip.add_address(naddr)
-    else:
+
+    if not has_valid_addr:
```
How about looking in the setting_ip  config and check if there are any addresses defined? (instead of using another flag)

EdDev commented on automation/Dockerfile:
```diff
@@ -15,6 +15,7 @@ RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == \
 
 RUN yum -y install iproute NetworkManager epel-release git cairo* && \
     yum -y install python2-pip gcc python-devel gobject-introspection-devel && \
+    yum -y install python2-ipaddress && \
```
It needs to be part of the nmstate installation, not part of the base env.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -332,3 +333,11 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+
+def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
+    for state in (desired_state, current_state):
+        for addr_dict in state.get('ipv6', {}).get('address', []):
```
I may miss something, but I think addr_dict is actually a list here. Please make sure you cover this flow here using tests.
BTW: Please do not use the name of the structure in the name. Use something like addr_data or address_state

cathay4t replied:
The `state['ipv6']['address'] a list of dict. So `addr_dict` is a dict here. I will rename it of course.
```json
"address": [
    {
        "ip": "fe80::8097:51a0:9ff4:c7a1",
        "prefix-length": 64
    }
],
```

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -332,3 +333,11 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
+    for state in (desired_state, current_state):
+        for addr_dict in state.get('ipv6', {}).get('address', []):
+            addr_str = '%s/%d' % (addr_dict['ip'], addr_dict['prefix-length'])
```
PLease use format()

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -341,3 +344,11 @@ def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
                 state['ipv6']['address'].remove(addr_dict)
+
+
+def _sort_ip_addresses(desired_state, current_state):
```
It will fit better outside this module, perhaps to an ipaddress module. We can delay this to an overall cleanup of netapplier.

cathay4t replied:
Will do that when we do cleanup in netappiler.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -341,3 +344,11 @@ def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
+def _sort_ip_addresses(desired_state, current_state):
+    for state in (desired_state, current_state):
+        for family in ('ipv4', 'ipv6'):
+            if state.get(family, {}).get('address', []):
```
How about using a variable as it is used again below and then do an inplace sort?

cathay4t replied:
Done.

cathay4t commented:
V3 patch set uploaded.

Changes since V2:
 * Changes based on feedback.
 * Unit test cased and integration test case are included now.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -333,3 +334,14 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+    for state in (desired_state, current_state):
+        for address in state.get('ipv6', {}).get('address', []):
+            addr_str = '{a[ip]}/{a[prefix-length]}'.format(a=address)
+            if nm.ipv6.is_link_local_addr(addr_str):
```
This seems to be non provider specific. Why have you placed it under `nm`?

cathay4t replied:
I normally don't like module dependency like:

`A.B.C depenps on `A.D`.

That's a other story. You shall have it place in libnmstate now.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -333,3 +334,14 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+        for address in state.get('ipv6', {}).get('address', []):
+            addr_str = '{a[ip]}/{a[prefix-length]}'.format(a=address)
+            if nm.ipv6.is_link_local_addr(addr_str):
+                state['ipv6']['address'].remove(address)
```
Removing inside the loop from the same list we iterate on feels dangerous in any language.
Perhaps collect the addresses needed for removal using a loop and just remove them.

cathay4t replied:
Fixed.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -333,3 +334,14 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+            if nm.ipv6.is_link_local_addr(addr_str):
+                state['ipv6']['address'].remove(address)
+        # Don't compare ipv6 status as link-local address might been removed.
+        if 'ipv6' in state and 'enabled' in state['ipv6']:
```
I do not understand this one... we do not compare ipv6 at all? So what is the cleanup all about?

cathay4t replied:
We do not compare the `state['ipv6']['state']` entry, we still compare the `state['ipv6']['address']`.

EdDev replied:
Oh, then if the user mentioned `disabled` and it is enabled with only the local-link, you want to pass verification?
That seems to hide reality: nmstate is unable to disable IPv6, why not express it?

BTW, a side question: What happens if we disable IPv6 using /proc/.. and then create a connection profile without the IPv6 setting? Will that still re-enable IPv6?

cathay4t replied:
NetworkManager will enable IPv6 when a profile got activated no match what previous proc state of that device is.

I don't think it's time to discuss that detail, we simple raise error and document that is not supported. We can come back to  that after we get a way to disable IPv6(via NetworkManager or disable IPv6 after NM profile activation).

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -46,15 +52,22 @@ def get_info(active_connection):
+            if is_link_local_addr(addr_str):
+                logging.warning('IPv6 link local address {} is ignored when '
+                                'applying desired state'.format(addr_str))
+                continue
```
I would prefer you use `else`, if possible, avoid `break` and `continue` in mid loop, it is bug prone (not always possible).

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -46,15 +52,22 @@ def get_info(active_connection):
 
 def create_setting(config):
     setting_ip = nmclient.NM.SettingIP6Config.new()
+    setting_ip.props.method = nmclient.NM.SETTING_IP6_CONFIG_METHOD_IGNORE
```
Why is this needed here?

cathay4t replied:
Without it, the unit test will fail, I haven't look into the detail yet. Will do.

cathay4t replied:
In unit test, the `settting_ip.props.address` will be set:
`<MagicMock name='NM.SettingIP6Config.new().props.addresses' id='139895920922704'>`
which is not empty list.
I have applied a workround in unit test for this.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -21,6 +21,7 @@
 import copy
 import six
 import time
+from operator import itemgetter
```
Please sort imports alphabetically.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -345,3 +347,11 @@ def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
+    for state in (desired_state, current_state):
+        for family in ('ipv4', 'ipv6'):
+            addrs = state.get(family, {}).get('address', [])
+            if addrs:
```
this is enough: addrs.sort(key=...)

EdDev commented:
I also suggest splitting this PR. I will help review and has a better chance to finalize things.

cathay4t commented:
> I also suggest splitting this PR. I will help review and has a better chance to finalize things.

We are talking about bug fixes in this PR, not new features, __all__ of the bugs should be fixed, splitting them into several PRs only waste my time.
I will do whatever code changes to make upstream happy for get them fixed.

EdDev commented:
> We are talking about bug fixes in this PR, not new features, all of the bugs should be fixed, splitting them into several PRs only waste my time.
> I will do whatever code changes to make upstream happy for get them fixed.

Could you elaborate on why it is more costly for you?
As with commits, making PR/s small has the advantage of getting them accepted faster (not delaying all due to one). Also, the reviewer can focus on one problem at a time, although that is possible by limiting the review window on a specific commit (this is what I usually do).
In gerrit (I like it more), every commit has a review independently, but github prefers comparing branches instead.

cathay4t commented:
V4 patch set uploaded.

Changes since V3:
 * `iplib` module moved to `libnmstate.iplib` and mimicking the ipaddress module in python3.7.
 * Test case code are included in the same patch with production code.
 * Unit test case for ip addresses internal ordering has been added.
 * Trivial fixes suggested above.

tyll commented on tests/integration/testlib/assertlib.py:
```diff
@@ -23,16 +23,38 @@
+    Just do simple string starts with check and prefix check, as this is purely
+    for test only.
+    """
+    return ip.startswith("fe80") and prefix >= 10
```
How about checking if it starts with `fe8`, `fe9`, `fea` or `feb` with this prefix size, then it should catch all link-local addresses?

cathay4t replied:
Good point. Thank you.

cathay4t commented:
V5 patch set uploaded.

Changes since V4:
 * Change `testlib` to support all IPv6 link-local prefix.
 * Fix type `adderss`.

EdDev commented on tests/lib/nm/ipv6_test.py:
```diff
@@ -29,6 +29,7 @@ def NM_mock():
 
 
 def test_create_setting_without_config(NM_mock):
+    NM_mock.SettingIP6Config.new().props.addresses = []
```
nit: Please add a newline. (separate setup, the tested action and the assert using newlines, so it will be a bit more clear which is which)

EdDev commented on tests/lib/netapplier_test.py:
```diff
@@ -92,6 +93,63 @@ def test_iface_admin_state_change(netinfo_nm_mock, netapplier_nm_mock):
+    netinfo_nm_mock.bond.is_bond_type_id.return_value = False
+    netinfo_nm_mock.ovs.is_ovs_bridge_type_id.return_value = False
+    netinfo_nm_mock.ovs.is_ovs_port_type_id.return_value = False
+    netinfo_nm_mock.ipv4.get_info.return_value = (
```
Then I think you should not inject a sorted version here, you need a non sorted one.

EdDev replied:
This test is not failing if I remove the production code sorting: https://github.com/nmstate/nmstate/pull/107/files#diff-6b531a9cd456f54d73ad106d72174513R327

EdDev commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -27,10 +27,21 @@
     eth1_desired_state = desired_state[INTERFACES][0]
     eth1_desired_state['state'] = 'up'
+    eth1_desired_state['ipv6'] = {'enabled': True}
+    # When a link came up, ipv6 link local address will be automatically
```
It seems to me like you use this to "fix" a valid scenario.
You also changed the tests: They started from a state=down, and now they start from state=up.
When the iface is down, ipv6 is enabled=False and when the iface is taken up, the user cannot (currently) leave the ipv6 disabled.
The minimalistic change needed is in the tests: to set ipv6 enabled (like we do wit the state=up).
A fixture can be used to provide the base dict (with state=up and ipv6.enabled=True), but not perform an action on the system. If you do that, I would do it in a separate PR or commit.

EdDev commented on tests/integration/static_ip_address_test.py:
```diff
@@ -61,6 +61,8 @@ def test_add_static_ipv4_with_full_state():
     eth1_desired_state['ipv4']['address'] = [
         {'ip': IPV4_ADDRESS3, 'prefix-length': 24}
     ]
+    # We don't have a way to disable IPv6 in libnmstate yet.
```
Use `FIXME:` or `TODO:` so we can revisit this.

EdDev replied:
Reminder to use `FIXME:` here as well.

EdDev commented on tests/integration/static_ip_address_test.py:
```diff
@@ -179,3 +181,70 @@ def test_add_iface_with_same_static_ipv4_address_to_existing(setup_eth1_ipv4):
+        # This sequence is intentionally made for IP address sorting.
+        {'ip': '2001:db8:1::1', 'prefix-length': 64},
+    ]
+    netapplier.apply(copy.deepcopy(desired_state))
```
not: For readability, please separate between setup, production call/execution and assertion (and teardown in needed) using newlines.
* For all tests added.

EdDev replied:
You did it everywhere except this line :)

EdDev commented on tests/integration/static_ip_address_test.py:
```diff
@@ -179,3 +181,70 @@ def test_add_iface_with_same_static_ipv4_address_to_existing(setup_eth1_ipv4):
+    eth1_desired_state['state'] = 'up'
+    eth1_desired_state['ipv6']['enabled'] = True
+    eth1_desired_state['ipv6']['address'] = [
+        {'ip': '2001:db8:2::1', 'prefix-length': 64},
```
Please use constants for (all) the addresses and prefixes. It is safer, especially when the same is asserted against later on.

EdDev commented on tests/integration/static_ip_address_test.py:
```diff
@@ -179,3 +181,70 @@ def test_add_iface_with_same_static_ipv4_address_to_existing(setup_eth1_ipv4):
+    # Make sure the link local address got ignored.
+    cur_state = statelib.show_only(('eth1',))
+    eth1_cur_state = cur_state[INTERFACES][0]
+    assert ({'ip': 'fe80::1', 'prefix-length': 64} not in
```
How about using eth1_desired_state['ipv6']['address'][0] ?

EdDev commented on tests/lib/netapplier_test.py:
```diff
@@ -92,6 +93,62 @@ def test_iface_admin_state_change(netinfo_nm_mock, netapplier_nm_mock):
+    config = {
+        'interfaces': [
+            {
+                "name": "foo",
```
single quotes please

cathay4t commented:
Thanks for the review.

I found the root cause of MTU test:
 * IPv6 require MTU < 1280.
 * When MTU change failed(like 0 or negative, etc), the IPv6 will be left disabled(Still investigating whether that's a race issue or not).

EdDev commented:
> When MTU change failed(like 0 or negative, etc), the IPv6 will be left disabled(Still investigating whether that's a race issue or not).

There are 3 expected failures there: zero, negative and too low mtu (40).
If we test now with `mtu=1000` and IPv6 gets enabled, it will also fail, right?

Still, I do not understand what race or error you see.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -314,6 +316,8 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
 
         iface_dstate, iface_cstate = _cleanup_iface_ethernet_state_sanitize(
             iface_dstate, iface_cstate)
+        _cleanup_iface_ipv6_link_local_addr(iface_dstate, iface_cstate)
```
Better follow Till recommendation as noted in commit f800a0e123a69b10b649f5d624bab79bea2ea0a5 (line 317 above).

cathay4t commented:
> > When MTU change failed(like 0 or negative, etc), the IPv6 will be left disabled(Still investigating whether that's a race issue or not).
> 
> There are 3 expected failures there: zero, negative and too low mtu (40).
> If we test now with `mtu=1000` and IPv6 gets enabled, it will also fail, right?
> 
> Still, I do not understand what race or error you see.

I was expect IPv6 been disabled for invalid MTU, but actually libnm revert back t o1500.

cathay4t commented:
V6 patch set uploaded.

Changes since V5:
 * Fix the MTU test with IPv6 enabled.
 * Trivial fixes based on feedback.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -334,3 +340,26 @@ def _cleanup_iface_ethernet_state_sanitize(desired_state, current_state):
+
+def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
+    for state in (desired_state, current_state):
+        if 'ipv6' not in state:
```
This is the opposite of cleanup, more like canonicalization. Another step?
And I think it can be done using:
```
desired_state = _dict_update({'ipv6': {'enabled': False, 'address': []}}, desired_state)
```
Do we need to do this for the current_state?

cathay4t replied:
Normally we don't do that for current state, but that require changes to existing unit test cases which does not have ipv6 states.
So I also canonicalize current state.

EdDev commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -43,6 +46,9 @@ def test_decrease_iface_mtu():
     eth1_desired_state['mtu'] = 1000
+    eth1_desired_state['ipv6']['enabled'] = False
+    eth1_desired_state['ipv6']['address'] = []
+    # IPv6 requires MTU >= 1280 (RFC 2460)
```
Can you add the comment before the ipv6 editing and explain why in this case we need to specify ipv6 `enabled=False`?
Is is also the default, so why even bother specifying it? I am missing something here.

Also, perhaps it is worth having now a test with `mtu=1280` for the "decrease" general case and the `mtu=1000` for the ipv6 case (only the latter needs explanation).

cathay4t replied:
It's kernel QE's job to test whether IPv6 got disabled when MTU < 1280, not us.

EdDev replied:
We need to check possible scenarios of input and see that nmstate behaves correctly.
Anyway, you have adjusted an existing test, so my main comment is about that. Adding another scenario is not a must, we can add it later.

EdDev commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -77,6 +85,8 @@ def test_decrease_to_zero_iface_mtu():
     eth1_desired_state['state'] = 'up'
     eth1_desired_state['mtu'] = 0
+    eth1_desired_state['ipv6']['enabled'] = True
+    # MTU set failure will revert back to 1500 MTU where IPv6 will be enabled.
```
I do not understand the comment. It reverts to the original value (default or not).
If nothing changed from the original test, I would drop these comments (line 104, 120 as well).

EdDev commented on tests/integration/testlib/assertlib.py:
```diff
@@ -29,5 +32,39 @@ def assert_state(desired_state_data):
     full_desired_state = statelib.State(current_state.state)
     full_desired_state.update(desired_state_data)
+    _ignore_ipv6_link_local(current_state.state, full_desired_state.state)
+    _sort_ip_addresses(current_state.state, full_desired_state.state)
```
You can add it now to State.normalize (after you rebase).

I do not know if the "ignore" one will fit as a method there (under normalize or other).

EdDev commented on tests/integration/testlib/assertlib.py:
```diff
@@ -23,16 +23,38 @@
 from .statelib import INTERFACES
 
 
+def _is_ipv6_link_local(ip, prefix):
```
Please place this func after `_ignore_ipv6_link_local`

EdDev commented on tests/lib/netapplier_test.py:
```diff
@@ -92,6 +93,63 @@ def test_iface_admin_state_change(netinfo_nm_mock, netapplier_nm_mock):
+    }
+    libnm_returned_config = copy.deepcopy(config)
+    for family in ('ipv4', 'ipv6'):
+        # In real world, libnm returned will sorted address in this case.
```
I do not think I understand (you need to massage the English a bit).
libnm will not sort its report about the addresses (at least, not per what I understand from the discussions with thaller on this)

cathay4t replied:
I don't know why thaller insist on that. On Fedora, the output of libnm is sorted. Example(I also included this example in code comment):

```bash
[fge@fed ~]$ nmstatectl show n0 |grep 192
                        "ip": "192.168.1.1",
                        "ip": "192.168.2.1",
[fge@fed ~]$ nmcli --get-values ipv4.addresses c show dummy-n0
192.168.2.1/24, 192.168.1.1/24
```

EdDev replied:
Based on your example here it sorted it in reverse order.
thaller mentioned it is adding it in the reverse order of how it is specified for adding.

But I think the main point is commented on line 145: If you sort it at the report step, the production code sorting is not checked, if it is broken (or get broken in the future), this test will not detect it.

As the netapplier module is so big, I am not even sure it does reach the sorting part with the proper values. This test is supposed to fail when you remove the verification ipv6 addresses sorting portion.. does it?

cathay4t replied:
This test is just simulate the failure we got when not sorting before comparing:
 * User input: `[192.168.2.1/24, 192.168.1.1/24]`
 * `netinfo.show()` got `[192.168.1.1/24, 192.168.2.1/24]`.

I don't get it why it does not detect the prolem.

EdDev commented on tests/lib/nm/ipv6_test.py:
```diff
@@ -117,3 +123,27 @@ def test_get_info_with_ipv6_config():
+    config = {
+        'enabled': True,
+        'address': [
+            {'ip': 'fe80::1', 'prefix-length': 64},
```
I would use constants so the reader will know which one is the link-local and which one is not, but it is not a must. Your call.

cathay4t commented:
V7 patch set uploaded.

Changes since V7:

 * Trivial fixes mentioned above.

EdDev commented on tests/integration/testlib/statelib.py:
```diff
@@ -97,6 +98,12 @@ def update(self, other_state):
 
     def normalize(self):
         self._sort_iface_lag_slaves()
+        for iface_state in self._state.get(INTERFACES, []):
```
It merits its own method: _ipv6_skeleton_canonicalization or something similar.

cathay4t commented:
V8 patch set uploaded.

Changes since V8:  
 * Changed the unit test of IP addresses sorting to non-mock version.
 * Created `_ipv6_skeleton_canonicalization()` to hold ipv6 subtree dict `setdefault()`

EdDev commented:
Thank you, looks great.
